### PR TITLE
Elimination of reference duplicates in the ReferenceManager should not depend on order

### LINKED ISSUE #5138: EE: Assembly unification of System.Numerics.Vectors.dll depends on order
1. Create C# Console Application target framework 4.6
2. Install NuGet package System.Numerics.Vectors 4.1.0
3. Add references to System.Numerics (4.0.0 from GAC) and System.Numerics.Vectors (4.1.0 from %userprofile%\.nuget\packages)
4. Change Program.cs to the code below
5. Set breakpoint at end of Main()
6. F5
7. Evaluate: ```new Vector<int>()```

```
using System.IO;
using System.Reflection;
using System.Numerics;

class Program
{
    static void Main(string[] args)
    {
        var v = new Vector2();
        F();
        G();
    } // break here

    static object F()
    {
        return new Vector<int>();
    }

    static void G()
    {
        Assembly.LoadFile(
            Path.Combine(
                Path.GetDirectoryName(typeof(object).Assembly.Location),
                "System.Numerics.Vectors.dll"));
    }
}
```

Result:
error CS1703: Multiple assemblies with equivalent identity have been imported: 'System.Numerics.Vectors.dll' and 'System.Numerics.Vectors.dll'. Remove one of the duplicate references.

Expected:
No errors. (No errors are reported if the order of F() and G() is reversed, and no errors are reported in either case with the legacy EE.)

### PR DESCRIPTION
**Pull request by tmat**
Fixes #5138

### CODE REVIEW

**Comment by tmat** [comment]
@cston @davkean @VSadov @AlekseyTs @agocke

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs [hunk 1]
```diff
@@ -2189,5 +2190,33 @@ public void ReusingCorLibManager()
             Assert.Same(assembly2.CorLibrary, assembly2);
             Assert.True(corlib1.ReferenceManagerEquals(corlib2));
         }
+
+        [WorkItem(43)]
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: 5138?

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs [hunk 2]
```diff
@@ -584,8 +584,13 @@ private static void AddModule(PEModule module, int referenceIndex, ResolvedRefer
             {
                 foreach (var other in sameSimpleNameIdentities)
                 {
-                    // only compare strong with strong (weak is never equivalent to strong and vice versa)
-                    if (other.Identity.IsStrongName && IdentityComparer.ReferenceMatchesDefinition(identity, other.Identity))
+                    // Only compare strong with strong (weak is never equivalent to strong and vice versa).
+                    // In order to eliminate duplicate references we need to try to match their identities in both directions since 
+                    // ReferenceMatchesDefinition is not neccessarily symmetric.
+                    // (e.g. System.Numerics.Vectors, Version=4.1+ matches System.Numerics.Vectors, Version=4.0, but the otherway around.)
```
Comment (cston) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs]: "... but _not_ the other way around."?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs [hunk 3]
```diff
@@ -2189,5 +2190,33 @@ public void ReusingCorLibManager()
...
+                options: TestOptions.ReleaseDll.WithCryptoPublicKey(TestResources.TestKeys.PublicKey_b03f5f7f11d50a3a),
+                assemblyName: "System.Numerics.Vectors");
+
+            Assert.Equal("System.Numerics.Vectors, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a", vectors41.Assembly.Identity.GetDisplayName());
+
+            var refVectors40 = vectors40.EmitToImageReference();
+            var refVectors41 = vectors41.EmitToImageReference();
+
+            var c1 = CreateCompilationWithMscorlib("", new[] { refVectors40, refVectors41 }, options: TestOptions.ReleaseDll.WithAssemblyIdentityComparer(DesktopAssemblyIdentityComparer.Default));
+            c1.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: Consider asserting result of unification

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs [hunk 4]
```diff
@@ -2189,5 +2190,33 @@ public void ReusingCorLibManager()
...
+
+            var c2 = CreateCompilationWithMscorlib("", new[] { refVectors41, refVectors40 }, options: TestOptions.ReleaseDll.WithAssemblyIdentityComparer(DesktopAssemblyIdentityComparer.Default));
+            c2.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: Here too.

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: BTW, what errors are reported without the fix?

Comment (tmat) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: The first one didn't report any errors, the second one reported duplicate references with non-equivalent identity.

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs (hunk 1 above)
Comment (tmat) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: Fixed.

File: src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Resolution.cs]: Fixed.

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs (hunk 3 above)
Comment (tmat) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: Done.

File: src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs (hunk 4 above)
Comment (tmat) [inline src/Compilers/CSharp/Test/Symbol/Compilation/ReferenceManagerTests.cs]: Done.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by tmat** [comment]
test linux please

**Pull request merged**